sgkit/stats/ibs.py

```diff
@@ -0,0 +1,260 @@
+from typing import Hashable
+
+import dask.array as da
+from numba import guvectorize
+from xarray import Dataset
+
+from sgkit import variables
+from sgkit.stats.aggregation import infer_call_allele_frequencies
+from sgkit.typing import ArrayLike
+from sgkit.utils import (
+    conditional_merge_datasets,
+    create_dataset,
+    define_variable_if_absent,
+)
+
+
+@guvectorize(  # type: ignore
+    [
+        "void(float32[:,:], float32[:,:], float32[:,:])",
+        "void(float64[:,:], float64[:,:], float64[:,:])",
+    ],
+    "(x,a),(y,a)->(x,y)",
+    nopython=True,
+    cache=True,
+)
+def identity_by_state(
+    af_x: ArrayLike, af_y: ArrayLike, out: ArrayLike
+) -> None:  # pragma: no cover
+    """Generalized U-function for computing the probability that
+    an allele drawn at random from a pair of individuals at a single
+    locus are identical in state.
+
+    This implementation computes identity in state probabilities
+    between all pairings of two sets of samples.
+
+    Parameters
+    ----------
+    af_x
+        Genotype allele counts of the first set of individuals with
+        shape (samples_x, alleles).
+    af_y
+        Genotype allele counts of the second set of individuals with
+        shape (samples_y, alleles).
+
+    Returns
+    -------
+    p_ibs
+        A matrix of identity by state probabilities between two sets
+        of samples with shape (samples_x, samples_y)
+
+    Notes
+    -----
+    This function is equivalent to
+    ``(af_x[..., None, :, :] * af_y[..., :, None, :]).sum(axis=-1)``.
+    """
+    x, n = af_x.shape
+    y, m = af_y.shape
+    assert n == m
+    for i in range(x):
+        for j in range(y):
+            ibs = 0.0
+            for a in range(n):
+                ibs += af_x[i, a] * af_y[j, a]
+            out[i, j] = ibs
+    return
+
+
+def pairwise_ibs(
```

Comment:
Would this function fit into the existing frameworks for pairwise distances? It's sort-of the same thing, isn't it?

Reply:
Good question. `pairwise_ibs` is currently calculating an "distance" matrix per variant (variants, samples, alleles) -> (variants, samples, samples). `pairwise_distance`  seems to only handle 2d matrices (M, N) -> (M, M) and doesn't vectorize along the additional axis (unless I'm missing something?) . 

`pairwise_ibs` should ideally be (variants, samples, alleles) -> (samples, samples), however I wanted to investigate a windowed version of this which is why I left the reduction until later.

I think it would be possible to implement `pairwise_ibs` as (variants, samples) -> (samples, samples) using the `pairwise_distance` framework however this would only handle biallelic dosages and I would like to handle the multi-allelic case.

Reply:
Ah, thanks for clearing up @timothymillar. I think what you have currently is good, there's no point in trying to shoehorn it into the pairwise distance stuff, I just wanted to make sure we weren't duplicating code or ideas.